### PR TITLE
[InstCombine] Canonicalize fcmp with inf

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch canonicalizes floating-point comparisons with inf:
```
fcmp olt X, +inf -> fcmp one X, +inf
fcmp ole X, +inf -> fcmp ord X, 0
fcmp ogt X, +inf -> false
fcmp oge X, +inf -> fcmp oeq X, +inf
fcmp ult X, +inf -> fcmp une X, +inf
fcmp ule X, +inf -> true
fcmp ugt X, +inf -> fcmp uno X, 0
fcmp uge X, +inf -> fcmp ueq X, +inf
fcmp olt X, -inf -> false
fcmp ole X, -inf -> fcmp oeq X, -inf
fcmp ogt X, -inf -> fcmp one X, -inf
fcmp oge X, -inf -> fcmp ord X, 0
fcmp ult X, -inf -> fcmp uno X, 0
fcmp ule X, -inf -> fcmp ueq X, -inf
fcmp ugt X, -inf -> fcmp une X, -inf
fcmp uge X, -inf -> true
```
Alive2: https://alive2.llvm.org/ce/z/FRqqDg

The motivation of this patch is to fix the regression found in https://github.com/dtcxzyw/llvm-opt-benchmark/pull/199#discussion_r1480974120.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -7751,6 +7751,48 @@ Instruction *InstCombinerImpl::visitFCmpInst(FCmpInst &I) {
...
+  // fcmp ult X, -inf -> fcmp uno X, 0
+  // fcmp ule X, -inf -> fcmp ueq X, -inf
+  // fcmp ugt X, -inf -> fcmp une X, -inf
+  // fcmp uge X, -inf -> true
+  const APFloat *C;
+  if (match(Op1, m_APFloat(C)) && C->isInfinity()) {
+    switch (C->isNegative() ? FCmpInst::getSwappedPredicate(Pred) : Pred) {
+    default:
+      break;
+    case FCmpInst::FCMP_OGT:
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7776]: so should ORD, UNO, TRUE and FALSE?

File: llvm/test/Transforms/InstCombine/canonicalize-fcmp-inf.ll [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+define i1 @olt_one(half %x) {
+; CHECK-LABEL: define i1 @olt_one(
+; CHECK-SAME: half [[X:%.*]]) {
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp olt half [[X]], 0xH3C00
+; CHECK-NEXT:    ret i1 [[CMP]]
+;
+  %cmp = fcmp olt half %x, 1.0
+  ret i1 %cmp
+}
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/canonicalize-fcmp-inf.ll:200]: test true / false for good measure. Also a few vector cases

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7778]: Done.

File: llvm/test/Transforms/InstCombine/canonicalize-fcmp-inf.ll [hunk 3]
```diff
@@ -177,6 +177,16 @@ define i1 @oge_pinf_fmf(half %x) {
   ret i1 %cmp
 }
 
+define <2 x i1> @olt_pinf_vec(<2 x half> %x) {
+; CHECK-LABEL: define <2 x i1> @olt_pinf_vec(
+; CHECK-SAME: <2 x half> [[X:%.*]]) {
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp one <2 x half> [[X]], <half 0xH7C00, half 0xH7C00>
+; CHECK-NEXT:    ret <2 x i1> [[CMP]]
+;
+  %cmp = fcmp olt <2 x half> %x, <half 0xH7c00, half 0xH7c00>
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/canonicalize-fcmp-inf.ll:186]: -inf too

**Review by arsenm — APPROVED** [review]

File: llvm/test/Transforms/InstCombine/canonicalize-fcmp-inf.ll (hunk 3 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/canonicalize-fcmp-inf.ll:186]: Done.

**Comment by dtcxzyw** [comment]
project/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7769:13: error: 16 enumeration values not handled in switch: 'FCMP_OEQ', 'FCMP_ONE', 'FCMP_UEQ'... [-Werror,-Wswitch]

**Comment by dtcxzyw** [comment]
> project/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7769:13: error: 16 enumeration values not handled in switch: 'FCMP_OEQ', 'FCMP_ONE', 'FCMP_UEQ'... [-Werror,-Wswitch]

Fixed by 934ba0d59e650644fe69462935982318b265ad2c.

**Pull request merged**